// @ncwidgets/reorder/dist/components/modal.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const styled_1 = __importDefault(require("@emotion/styled"));
const StyledHeader = styled_1.default.h1 `
  font-size: 1.25rem;
  line-height: 1.3;
  margin-bottom: 1rem;
`;
const StyledModal = styled_1.default.div `
  position: absolute;
  width: 50%;
  padding: 1rem 1rem 1.25rem;
  text-align: center;
  top: 50%;
  left: 50%;
  transform: translate(-50%,-50%);
  background-color: #fff;
  border-radius: 0.25rem;
  box-shadow: 0 6px 12px 0 rgba(0,0,0,0.2);

  button {
    border: 1px solid #dfdfe3;
    color: rgb(23, 162, 184);
    padding: 0.75em 1.25em;
    background: #fff;
    border-radius: 6px;
  }

  button:hover {
    border: 1px solid rgb(23, 162, 184);
    background: rgb(23, 162, 184);
    color: #fff;
  }
`;
const StyledOverlay = styled_1.default.div `
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: rgba(223, 223, 227, 0.8);
  z-index: 100;
`;
const getModalContent = ({ modified, collection }) => {
    let title = '';
    let action = '';
    if (modified === 'unset') {
        title = `Order for collection '${collection}' hasn\'t been set yet`;
        action = 'Start Ordering';
    }
    if (modified === 'modified') {
        title = `Collection '${collection}' have been changed`;
        action = 'Apply changes';
    }
    return { title, action };
};
exports.Modal = ({ collection, modified, handleDisplayChange }) => {
    const { title, action } = getModalContent({ modified, collection });
    return (React.createElement(StyledOverlay, null,
        React.createElement(StyledModal, null,
            React.createElement(StyledHeader, null, title),
            React.createElement("button", { onClick: () => handleDisplayChange() }, action))));
};
const StyledEmptyMessage = styled_1.default.div `
  padding: 1rem;
  text-align: center;
`;
exports.EmptyMessage = ({ className, collection }) => (React.createElement(StyledEmptyMessage, { className: className },
    React.createElement(StyledHeader, null,
        "Collection '",
        collection,
        "' is empty"),
    React.createElement("p", null,
        React.createElement("a", { href: `#/collections/${collection}/new`, target: "blank" }, "Create new entries"))));
