// @ncwidgets/reorder/dist/notice.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const wrapperStyle = {
    position: 'relative',
    width: '100%',
    padding: '1rem 1rem 0 1rem',
    boxSizing: 'border-box',
    backgroundColor: '#dfdfe3',
    borderTopRightRadius: '3px',
    marginBottom: '-5px',
};
const style = {
    position: 'relative',
    width: '100%',
    padding: '0.25rem',
    textAlign: 'center',
    color: '#8C5708',
    borderRadius: '3px',
    lineHeight: '1.5',
    backgroundColor: '#FFC23E',
};
exports.NoticeBar = ({ children, displayChange }) => (React.createElement("div", { style: wrapperStyle },
    React.createElement("div", { style: style },
        children,
        ' ',
        React.createElement("button", { onClick: () => displayChange() }, "Show"))));
